### PR TITLE
Integrate RMM

### PR DESCRIPTION
**Pull request by mzient**
#### Why we need this PR?
*Pick one, remove the rest*
- Refactoring to use RAPIDS memory manager

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     * Use RMM fork with interfaces we need
     * Replace memory_resource and stream_memory_resource with interfaces from RMM
     * Adjust code to acommodate the interfaces
     * Add RMM dependencies to build system
     * Add RMM and ints dependencies to include paths
 - Affected modules and functionalities:
     * build
     * core/memory manager
 - Key points relevant for the review:
     * N/A
 - Validation and testing:
     * N/A
 - Documentation (including examples):
     * N/A

**JIRA TASK**: [DALI-1782] [DALI-1783]

### CODE REVIEW

**Comment by mzient** [comment]
!build

File: cmake/Dependencies.common.cmake [hunk 1]
```diff
@@ -199,3 +199,19 @@ add_library(cocoapi STATIC ${SOURCE_FILES})
 set_target_properties(cocoapi PROPERTIES POSITION_INDEPENDENT_CODE ON)
 list(APPEND DALI_LIBS cocoapi)
 list(APPEND DALI_EXCLUDES libcocoapi.a)
+
+##################################################################
+# RAPIDS Memory Manager
+##################################################################
+include_directories(SYSTEM ${PROJECT_SOURCE_DIR}/third_party/rmm/include)
```
Comment (JanuszL) [inline cmake/Dependencies.common.cmake:206]: Can you check if that compiler for QNX, aarch64 and SBSA?

Comment (mzient) [reply cmake/Dependencies.common.cmake:206]: QNX passed.

File: dali/core/mm/mm_test_utils.h [hunk 2]
```diff
@@ -29,18 +30,24 @@ namespace mm {
...
       throw std::bad_alloc();
     void *ret;
     if (security_check) {
-      ret = upstream_->allocate(extra..., bytes + sizeof(sentinel_t), alignment);
-      detail::write_sentinel<sentinel_t>(ret, bytes);
+      ret = ua(bytes + sizeof(sentinel_t), alignment);
+      detail::write_sentinel<sentinel_t>(ret, bytes, std::forward<Extra>(extra)...);
     } else {
-      ret = upstream_->allocate(extra..., bytes, alignment);
+      ret = upstream_->allocate(bytes, alignment, std::forward<Extra>(extra)...);
```
Comment (JanuszL) [inline dali/core/mm/mm_test_utils.h:50]: Why use `ua` in one place and `upstream_` directly in the second?

File: dali/core/mm/mm_test_utils.h [hunk 3]
```diff
@@ -29,18 +30,24 @@ namespace mm {
 namespace test {
 
 template <bool owning, bool security_check,
-          typename Base, typename Upstream, typename... ExtraParams>
-class test_resource_wrapper : public Base {
+          typename Base, typename Upstream>
+class test_resource_wrapper;
+
+template <bool owning, bool security_check, typename Upstream>
+class test_resource_wrapper_impl {
```
Comment (JanuszL) [inline dali/core/mm/mm_test_utils.h:37]: Why you need to split wrapper into impl and wrapper itself?

File: dali/core/mm/mm_test_utils.h (hunk 2 above)
Comment (mzient) [reply dali/core/mm/mm_test_utils.h:50]: It's a bug. It should be `ua`.

File: dali/core/mm/monotonic_resource_test.cc [hunk 4]
```diff
@@ -69,25 +69,26 @@ TEST(MMTest, MonotonicHostResource) {
 }
 
 TEST(MMTest, MonotonicDeviceResource) {
-  test_host_resource upstream;
+  test_device_resource upstream;
```
Comment (JanuszL) [inline dali/core/mm/monotonic_resource_test.cc:72]: So RMM is not applicable to the host memory yet, or there is another reason for this change?

File: include/dali/core/mm/monotonic_resource.h [hunk 5]
```diff
@@ -260,6 +272,10 @@ class monotonic_device_resource : public memory_resource {
   SmallVector<upstream_block, 8> blocks_;
 };
 
+using monotonic_host_resource = monotonic_memory_resource<memory_kind::host>;
```
Comment (JanuszL) [inline include/dali/core/mm/monotonic_resource.h:275]: Maybe I'm missing something, but order doesn't have any default value nor is provided in this using statement.

Comment (mzient) [reply include/dali/core/mm/monotonic_resource.h:275]: The default argument is defined in line 84.

File: dali/core/mm/monotonic_resource_test.cc (hunk 4 above)
Comment (mzient) [reply dali/core/mm/monotonic_resource_test.cc:72]: Previously the resources did not distinguish memory type.

File: dali/core/mm/mm_test_utils.h (hunk 3 above)
Comment (mzient) [reply dali/core/mm/mm_test_utils.h:39]: The `impl` has template callable UpstreamAlloc and UpstreamDealloc which can resolve to `upstream_->(de)allocate` or `upstream_->(de)allocate_async` (with stream). Without that I'd need to have 2 or 3 near-copies of alloc/dealloc functions.

Comment (JanuszL) [reply dali/core/mm/mm_test_utils.h:39]: ok

File: dali/core/mm/monotonic_resource_test.cc (hunk 4 above)
Comment (JanuszL) [reply dali/core/mm/monotonic_resource_test.cc:72]: Understood

File: include/dali/core/mm/monotonic_resource.h (hunk 5 above)
Comment (JanuszL) [reply include/dali/core/mm/monotonic_resource.h:269]: Missed that.

**Review by jantonguirao — APPROVED** [review]

**Review by JanuszL — APPROVED** [review]

**Pull request merged**